### ISSUE TITLE
Error (time out) when building the project on macOS

### ISSUE THREAD

**Issue by 0xced** [aspnet/AspNetCore#4605]
I can't get the AspNetCore project to build. Here is the output of running `./build.sh --verbose` on master (macOS Sierra 10.12.6):
```
./build.sh --verbose
Using KoreBuild 3.0.0-build-20181206.3
Using cached .NET Framework reference assemblies from /Users/0xced/.dotnet/buildtools/netfx/4.7.2
warning: dotnet found on the system PATH is '/usr/local/share/dotnet/dotnet' but KoreBuild will use '/Users/0xced/.dotnet/dotnet'.
warning: Adding '/Users/0xced/.dotnet' to system PATH permanently may be required for applications like Visual Studio for Mac or VS Code to work correctly.
debug  : Installing .NET Core SDK 3.0.100-preview-009750
.NET Core SDK 3.0.100-preview-009750 is already installed. Skipping installation.
debug  : Building /Users/0xced/Projects/AspNetCore
debug  : Enabling binary logging
debug  : dotnet = /Users/0xced/.dotnet/dotnet
debug  : Noop = false
>>> dotnet publish /Users/0xced/Projects/AspNetCore/build/tasks/RepoTasks.csproj --configuration Release --output /Users/0xced/Projects/AspNetCore/build/tasks/bin/publish/ -nologo -p:RepoTasksSdkPath=/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/scripts/../msbuild/KoreBuild.RepoTasks.Sdk/Sdk/ @/Users/0xced/Projects/AspNetCore/artifacts/logs/msbuild.logger.rsp
  Restore completed in 27.08 ms for /Users/0xced/Projects/AspNetCore/build/tasks/RepoTasks.csproj.
  RepoTasks -> /Users/0xced/Projects/AspNetCore/build/tasks/bin/Release/netcoreapp2.1/RepoTasks.dll
  RepoTasks -> /Users/0xced/Projects/AspNetCore/build/tasks/bin/publish/

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.65
<<< dotnet [0]
debug  : Invoking msbuild with '/m
/nodeReuse:false
/p:KoreBuildVersion=3.0.0-build-20181206.3
/p:SuppressNETCoreSdkPreviewMessage=true
/p:RepositoryRoot="/Users/0xced/Projects/AspNetCore/"
"/bl:/Users/0xced/Projects/AspNetCore/artifacts/logs/msbuild.binlog"
"/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/scripts/../KoreBuild.proj"'
>>> dotnet msbuild @/Users/0xced/Projects/AspNetCore/artifacts/logs/msbuild.rsp @/Users/0xced/Projects/AspNetCore/artifacts/logs/msbuild.logger.rsp
Microsoft (R) Build Engine version 15.9.19+g938f3292a0 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

/Users/0xced/.dotnet/sdk/3.0.100-preview-009750/MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/Users/0xced/.dotnet/sdk/3.0.100-preview-009750/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/Users/0xced/.dotnet/sdk/3.0.100-preview-009750/dotnet.dll -maxcpucount /m -verbosity:m /bl:/Users/0xced/Projects/AspNetCore/artifacts/logs/msbuild.binlog /clp:Summary /nodeReuse:false /p:KoreBuildVersion=3.0.0-build-20181206.3 /p:SuppressNETCoreSdkPreviewMessage=true /p:RepositoryRoot=/Users/0xced/Projects/AspNetCore/ /Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/scripts/../KoreBuild.proj
  Restoring NPM modules
  npm WARN FunctionalTests No description
  npm WARN FunctionalTests No repository field.
  npm WARN FunctionalTests No license field.
  
  audited 17652 packages in 12.291s
  found 0 vulnerabilities
  
  Installing .NET Core Runtime (x64) 3.0.0-preview-27122-01
  dotnet-install output:
  dotnet-install: Calling: machine_has curl
  dotnet-install: Calling: calculate_vars 
  dotnet-install: Calling: get_normalized_architecture_from_architecture x64
  dotnet-install: normalized_architecture=x64
  dotnet-install: Calling: get_specific_version_from_version https://dotnetcli.blob.core.windows.net/dotnet LTS x64 3.0.0-preview-27122-01
  dotnet-install: specific_version=3.0.0-preview-27122-01
  dotnet-install: Calling: construct_download_link https://dotnetcli.blob.core.windows.net/dotnet LTS x64 3.0.0-preview-27122-01
  dotnet-install: Calling: get_current_os_name 
  dotnet-install: download_link=https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz
  dotnet-install: Calling: construct_legacy_download_link https://dotnetcli.blob.core.windows.net/dotnet LTS x64 3.0.0-preview-27122-01
  dotnet-install: Calling: get_legacy_os_name 
  dotnet-install: legacy_download_link=https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-osx-x64.3.0.0-preview-27122-01.tar.gz
  dotnet-install: Calling: resolve_installation_path /Users/0xced/.dotnet
  dotnet-install: install_root=/Users/0xced/.dotnet
  dotnet-install: Calling: check_pre_reqs 
  dotnet-install: Calling: install_dotnet 
  dotnet-install: Calling: is_dotnet_package_installed /Users/0xced/.dotnet shared/Microsoft.NETCore.App 3.0.0-preview-27122-01
  dotnet-install: Calling: combine_paths /Users/0xced/.dotnet shared/Microsoft.NETCore.App
  dotnet-install: combine_paths: root_path=/Users/0xced/.dotnet
  dotnet-install: combine_paths: child_path=shared/Microsoft.NETCore.App
  dotnet-install: Calling: combine_paths /Users/0xced/.dotnet/shared/Microsoft.NETCore.App 3.0.0-preview-27122-01
  dotnet-install: combine_paths: root_path=/Users/0xced/.dotnet/shared/Microsoft.NETCore.App
  dotnet-install: combine_paths: child_path=3.0.0-preview-27122-01
  dotnet-install: is_dotnet_package_installed: dotnet_package_path=/Users/0xced/.dotnet/shared/Microsoft.NETCore.App/3.0.0-preview-27122-01
  dotnet-install: Zip path: /var/folders/1l/rbhqr1jx4fdbgxmf58vwdw700000gq/T//dotnet.jRE126vpV
  dotnet-install: Downloading link: https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz
  dotnet-install: Calling: download https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz /var/folders/1l/rbhqr1jx4fdbgxmf58vwdw700000gq/T//dotnet.jRE126vpV
  dotnet-install: Calling: machine_has curl
  dotnet-install: Calling: downloadcurl https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz /var/folders/1l/rbhqr1jx4fdbgxmf58vwdw700000gq/T//dotnet.jRE126vpV
/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/modules/KoreBuild.Tasks/module.targets(44,5): error : dotnet-install of .NET Core Runtime (x64) 3.0.0-preview-27122-01 timed out after 240 seconds. [/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/KoreBuild.proj]

Build FAILED.

/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/modules/KoreBuild.Tasks/module.targets(44,5): error : dotnet-install of .NET Core Runtime (x64) 3.0.0-preview-27122-01 timed out after 240 seconds. [/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/KoreBuild.proj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:04:14.26
error  : dotnet failed with exit code 1
```

<details>
 <summary> And here is my  ~/.dotnet/dotnet --info</summary>

```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview-009750
 Commit:    84dcc2bfd8

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.12
 OS Platform: Darwin
 RID:         osx.10.12-x64
 Base Path:   /Users/0xced/.dotnet/sdk/3.0.100-preview-009750/

Host (useful for support):
  Version: 3.0.0-preview-27113-06
  Commit:  48270a44c2

.NET Core SDKs installed:
  2.1.302 [/Users/0xced/.dotnet/sdk]
  2.1.400 [/Users/0xced/.dotnet/sdk]
  3.0.100-preview-009750 [/Users/0xced/.dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.2 [/Users/0xced/.dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 3.0.0-alpha1-10663 [/Users/0xced/.dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.2 [/Users/0xced/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-alpha1-10663 [/Users/0xced/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.9 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0-preview1-26618-02 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0-preview2-26905-02 [/Users/0xced/.dotnet/shared/Microsoft.NETCore.App]
```
</details>

What's that timeout? Is there a way I could enable even more verbose output to understand what's going on?

**Comment by natemcmaster** [comment]
The logs indicate that the issue was a timeout downloading https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz.
```
  dotnet-install: Calling: downloadcurl https://dotnetcli.blob.core.windows.net/dotnet/Runtime/3.0.0-preview-27122-01/dotnet-runtime-3.0.0-preview-27122-01-osx-x64.tar.gz /var/folders/1l/rbhqr1jx4fdbgxmf58vwdw700000gq/T//dotnet.jRE126vpV
/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/modules/KoreBuild.Tasks/module.targets(44,5): error : dotnet-install of .NET Core Runtime (x64) 3.0.0-preview-27122-01 timed out after 240 seconds. [/Users/0xced/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/KoreBuild.proj]
```

Try running again. Sometimes Azure storage is flaky.

Side note: what's your internet speed? Most of us are on relatively fast internet, so our timeouts are often tuned to what we perceive as 'slow.'

**Comment by 0xced** [comment]
I already tried about ten times and it always fail with the timeout. Indeed, the download is **very** slow, I can see `curl` downloading between 50 and 100 KB/s which is way under my usual download speed.

So here is what I did: I looked at `~/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/modules/KoreBuild.Tasks/module.targets(44,5)` and saw that it calls `~/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/scripts/dotnet-install.sh`. After some trial and error I guessed the parameters to pass to this script so that it performs the same installation that is initiated by MSBuild:

```
$ cd ~/.dotnet/buildtools/korebuild/3.0.0-build-20181206.3/scripts
$ ./dotnet-install.sh --verbose --no-cdn --runtime dotnet -v 3.0.0-preview-27122-01
```

And it worked! The installation of .NET Core Runtime (x64) 3.0.0-preview-27122-01 succeeded. Then I came back to the AspNetCore repository and ran `./build.sh --verbose` again. It detected that the runtime was now installed and went on with the build, yay!

Of course this whole process would be much easier if MSBuild did not impose an arbitrary 240 seconds timeout. 😞
